juju/juju
# Split logger for charms

The logger in the charmhub and charms facade packages require a
cyclic dependency on the interface. This causes lots of problems
when implementing the logger. To make matters worse, the interface
also had a loggo.Logger attached to some loggers, which makes it
impossible to satisfy that interface if you've got a sub-type of that
interface.

The solution is to make two interfaces a logger factory and a logger.
This removes the cyclic nature of the interface and cleans up what
the actual dependency is.

This is a follow on from PRs: 
 - https://github.com/juju/juju/pull/16672
 - https://github.com/juju/juju/pull/16681
 - https://github.com/juju/juju/pull/16683

## Checklist


## QA steps

```sh
$ juju bootstrap lxd test --build-agent
```

In a different terminal, run `juju debug-log -m controller`

Then run the following to see it all still works.

```sh
$ juju model-config -m controller logging-config="<root>=INFO;#charmhub=TRACE;#http=TRACE"
$ juju add-model default
$ juju deploy ubuntu
```

## Links

**Jira card:** JUJU-5137
